File: InstagramRemakeJava/app/src/main/java/com/silvandrade/instagramremakejava/login/presentation/LoginPresenter.java
package com.silvandrade.instagramremakejava.login.presentation;

import android.os.Handler;

import com.silvandrade.instagramremakejava.R;
import com.silvandrade.instagramremakejava.common.model.UserAuth;
import com.silvandrade.instagramremakejava.common.presenter.Presenter;
import com.silvandrade.instagramremakejava.common.util.Strings;
import com.silvandrade.instagramremakejava.login.datasource.LoginDataSource;


public class LoginPresenter implements Presenter<UserAuth> {

    private final LoginView view;
    private final LoginDataSource dataSource;

    public LoginPresenter(LoginView view, LoginDataSource dataSource) {
        this.view = view;
        this.dataSource = dataSource;
    }

    public void login(String email, String password) {

        if (!Strings.emailIsValid(email)) {
            view.onFailureForm(view.getContext().getString(R.string.invalid_email), null);
            return;
        }

        view.showProgressBar();
        dataSource.login(email, password, this);

        new Handler().postDelayed(() -> {
            view.hideProgressBar();
        }, 2000);
    }

    @Override
    public void onSuccess(UserAuth userAuth) {
        view.onUserLogged();
    }

    @Override
    public void onError(String message) {
        view.onFailureForm(null, message);
    }

    @Override
    public void onComplete() {
        view.hideProgressBar();
    }
}

File: Instagram/app/src/main/java/co/tiagoaguiar/course/instagram/login/presenter/LoginPresenter.kt
package co.tiagoaguiar.course.instagram.login.presenter

import android.util.Log
import android.util.Patterns
import co.tiagoaguiar.course.instagram.R
import co.tiagoaguiar.course.instagram.common.model.UserAuth
import co.tiagoaguiar.course.instagram.login.Login
import co.tiagoaguiar.course.instagram.login.data.LoginCallback
import co.tiagoaguiar.course.instagram.login.data.LoginRepository

class LoginPresenter(
    private var view: Login.View?,
    private var repository: LoginRepository) : Login.Presenter {

    override fun login(email: String, password: String) {
        val isEmailValid: Boolean = Patterns.EMAIL_ADDRESS.matcher(email).matches()
        val isPasswordValid: Boolean = password.length >= 8;

        if(!isEmailValid) {
            view?.displayEmailFailure(R.string.invalid_email)
        } else {
            view?.displayEmailFailure(null)
        }

        if(!isPasswordValid) {
            view?.displayPasswordFailure(R.string.invalid_password)
        } else {
            view?.displayPasswordFailure(null)
        }

        if(isEmailValid && isPasswordValid) {
            view?.showProgress(true)

            repository.login(email, password, object: LoginCallback {
                override fun onSuccess(userAuth: UserAuth) { // Se o model devolver sucesso.
                    view?.onUserAuthenticated() // Informa a view que o usuário foi a autenticado
                }

                override fun onFailure(message: String) { // Se o model devolver falha.
                    view?.onUserUnauthorized(message) // Informa a view sobre o erro.
                }

                override fun onComplete() { // Sempre será executado.
                    view?.showProgress(false) // Informa a view para esconder a progress.
                }
            })
        }
    }

    override fun onDestroy() {
        view = null
    }
}